Implement Cupertino Sheet

jamesblasco opened the pull request:
<a href="https://jamesblasco.github.io/modal_bottom_sheet/#/"><img height="300px" src="https://github.com/jamesblasco/modal_bottom_sheet/blob/screenshots/preview.png?raw=true"></a>
## Description

Draft for bringing support to the Cupertino modal bottom sheet. I would like to have some feedback on the code in general before focusing on improving the quality, adding docs and tests.

It needs to be decided on how the current showModalBottomSheet coexists with showCustomModalBottomSheet, showMaterialBottomSheet, and showCupertinoModalBottomSheet.

And how to allow routes to define the forward animation of the previous route when the current one is being popped.

This code is from the [modal_bottom_sheet](https://pub.dev/packages/modal_bottom_sheet) package, with small differences that will be published in the next version of the package.

The API is similar for all the modals:
```dart
showCupertinoModalBottomSheet(
  context: context,
  builder: (context) => Container(),
)
```

Working example project with this PR -> https://github.com/jamesblasco/modal_bottom_sheet_example
Web demo with the package -> https://jamesblasco.github.io/modal_bottom_sheet/#/

## Related Issues

Closes https://github.com/flutter/flutter/issues/42560

## Tests

Clearly one of the most important works here would be to add test to this code.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

jamesblasco committed:
Add base project from https://github.com/jamesblasco/modal_bottom_sheet + null safety

DanielEdrisian commented:
Hey, thanks for submitting the PR! I'm going through this and I have a quick question:

How can I get the scaling effect with this? This is my result right now:

<img src="https://user-images.githubusercontent.com/11810973/95529779-b580f480-0990-11eb-8087-48f45b7459be.gif" width=250>

My flutter code:

```dart
CupertinoButton(
      child: Text('Show Flutter Modal'),
      onPressed: () {
        showCupertinoModalBottomSheet(
          context: context,
          builder: (context) => Container(),
        );
      },
    )
```

(What you see at first is the standard iOS modal done in SwiftUI. The second one is done in Flutter.)

jamesblasco commented:
Hello!

So far I only implemented it for routes that are `MaterialPageRoute`. I also should do it with `CupertinoPageRoute,` or all PageRoutes.

In the case of a CupertinoApp, it uses CupertinoPageRoute so now it doesn't work out of the box, and you should create the routes inside onGenerateRoutes. I will make a commit supporting it

<details>
<summary>Working demo code</summary>

```dart

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(platform: TargetPlatform.iOS),
      title: 'BottomSheet Modals',
      home: Builder(
        builder: (context) => Scaffold(
          body: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              ListTile(
                title: Text('Cupertino Sheet'),
                onTap: () {
                  showCupertinoModalBottomSheet(
                    expand: true,
                    context: context,
                    builder: (context) => Container(),
                  );
                },
              )
            ],
          ),
        ),
      ),
    );
  }
}
```

</details>


Note: In the package, I can not edit `MaterialPageRoute`. Devs always needed to replace it with `MaterialWithModalsPageRoute`. 

Now I created a mixin `AllowsRouteWithPreviousTransitionMixin` that out of the box allows Routes that contains the mixin `RouteWithPreviousTransitionMixin`. Every route that uses that last mixin will need to apply a method that defines the transition of the previous route. 

Eg. The root page doesn't transition when first display, but will scale back when a modal sheet route closes

jamesblasco committed:
Add support for CupertinoPageRoute

jamesblasco commented:
Now you should be able to use `MaterialPageRoute` or `CupertinoPageRoute`

DanielEdrisian commented:
Hey man, that's great! Will take a look.

You've already pointed out that this PR needs some tests before landing--perfect. We should also put in the effort to semi-extensively document the code (beyond the occasional comments.) [Here's an example of what I mean.](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/modal_barrier.dart)

After those two are done, I have a couple of questions about the curve function that I'd like to ask.

Thanks James!

jamesblasco commented:
Thanks for the response @DanielEdrisian

Before I started to improve and document the code, I would like to get some feedback about the things I pointed out in the PR description. I think they are important. I will comment on them in the code separately.

jamesblasco commented on packages/flutter/lib/src/widgets/modal_button_sheet_route.dart:
```diff
@@ -0,0 +1,321 @@
+
+// Route that allows the next route to define the animation transition of this route
+// when the route appears back after the next one is popped
+mixin AllowsRouteWithPreviousTransitionMixin<T> on PageRoute<T> {
```
Right now the back/forth transitions are defined by the route itself. For the modals, the previous route transition needs to be defined by the modal route itself. I am not sure if my approach here is the best way to do it, but this is the only way I managed to do it with the current situation of the framework.

It would be nice to add the modals as the new Page class for the Navigator 2.0

jamesblasco commented on packages/flutter/lib/src/material/modal_bottom_sheet.dart:
```diff
@@ -0,0 +1,124 @@
+import 'theme.dart';
+
+/// Shows a modal material design bottom sheet.
+Future<T> showMaterialModalBottomSheet<T>({
```
Should we replace the current showModalBottomSheet with this one? Or should I remove this one?

jamesblasco commented on packages/flutter/lib/src/cupertino/modal_bottom_sheet.dart:
```diff
@@ -0,0 +1,389 @@
+}
+
+//  This is can be removed once MaterialPageRoute suppors animated modals
+class CupertinoScaffold extends StatefulWidget {
```
This I think it should be removed

xster replied:
I would definitely not re-export something specifically for routes with such a generic name.

jamesblasco commented on packages/flutter/lib/src/widgets/modal_button_sheet.dart:
```diff
@@ -0,0 +1,652 @@
+  return PointerDeviceKind.unknown;
+}
+
+// Copied from bottom_sheet.dart as is a private class
+// https://github.com/flutter/flutter/issues/51627
```
This class `BottomSheetSuspendedCurve` I think should be made public, moved to another file, and be reused as it would be now three times duplicated.

jamesblasco commented on packages/flutter/lib/src/widgets/modal_button_sheet.dart:
```diff
@@ -0,0 +1,652 @@
+/// subtree inside a modal bottom sheet.
+///
+/// We want to remove this and use [PrimaryScrollController].
+/// This issue should be solved first https://github.com/flutter/flutter/issues/64236
```
I think this issue should be solved so Scaffold can inherit the `PrimaryScrollController` and `ModalScrollController` should not be needed.

jamesblasco commented on packages/flutter/lib/src/widgets/modal_button_sheet_route.dart:
```diff
@@ -0,0 +1,321 @@
+Future<T> showCustomModalBottomSheet<T>({
+  required BuildContext context,
+  required WidgetBuilder builder,
+  required WidgetWithChildBuilder containerWidget,
```
Not sure how to call this builder but don't like the current name

xster commented on packages/flutter/lib/src/cupertino/modal_bottom_sheet.dart:
```diff
@@ -0,0 +1,389 @@
+  }
+}
+
+class _CupertinoModalTransition extends StatelessWidget {
```
Looking at the lines above, is this meant for the z-level retracting perspective move? If so, it could use a more descriptive name.

jamesblasco replied:
Yes, It is the transition from the previous route that scales down. I am not sure how to call the transition.

xster commented on packages/flutter/lib/src/cupertino/modal_bottom_sheet.dart:
```diff
@@ -0,0 +1,389 @@
+    return _CupertinoScaffold(
+      animation: animationController,
+      topRadius: widget.topRadius,
+      child: _CupertinoModalTransition(
```
Not very clear what this group of classes is meant to solve. Could use some further description.

jamesblasco replied:
I deleted it. People were having problems changing the default MaterialPageRoute to MaterialWithModalPageRoute when using the package, so I created this widget `CupertinoScaffold` that would animate the same without needed to change the route. I think it is not needed anymore as the default routes would support the animation without any change needed

xster commented on packages/flutter/lib/src/widgets/modal_button_sheet.dart:
```diff
@@ -0,0 +1,652 @@
+import 'package:flutter/scheduler.dart';
+import 'package:flutter/widgets.dart';
+
+const Curve _decelerateEasing = Cubic(0.0, 0.0, 0.2, 1.0);
```
before reading this class, my first impression is that I'm tempted to think that we can just somewhat follow the https://github.com/flutter/flutter/blob/354e2a57b1ea20cda88ae1c24b5283dced54fde2/packages/flutter/lib/src/material/page_transitions_theme.dart#L561 example and not push something so visually opinionated in terms of implementation (such as curves, velocities etc) into the widgets layer. Having some general semantics of normal vs fullscreen vs modal sheet is ok but there's a lot of implementation details here. 

i.e. I have trouble imagining a case where someone wants to base a transition off of this one, not use material or cupertino and find these existing specifications in terms of velocity, duration etc useful.

jamesblasco replied:
Yeah, that was my approach from the beginning, but a lot of people have been asking for this level of customization.  The package has them default from the beginning and has been adding these params durations, curves by demand.

If `modal_bottom_sheet.dart` is merged with `modal_bottom_sheet_route.dart`, `ModalBottomSheet` class could be private

xster commented on packages/flutter/lib/src/widgets/modal_button_sheet.dart:
```diff
@@ -0,0 +1,652 @@
+
+/// A custom bottom sheet.
+///
+/// The [ModalBottomSheet] widget itself is rarely used directly. Instead, prefer to
```
Same as mentioned on top. I would perhaps not use the "bottom sheet" terminology since they already have a specific meaning.

xster commented on packages/flutter/lib/src/widgets/modal_button_sheet.dart:
```diff
@@ -0,0 +1,652 @@
+  VelocityTracker? _velocityTracker;
+  DateTime? _startTime;
+
+  void _handleScrollUpdate(ScrollNotification notification) {
```
having all these touch handling stuff is understandable but I think they can just be in the cupertino package. For prior art, see all the back gesture handling and how they're all re-used by the MaterialPageRoute as well through the PageTransitionsTheme.

xster commented on packages/flutter/lib/src/widgets/modal_button_sheet.dart:
```diff
@@ -0,0 +1,652 @@
+  }
+}
+
+class _ModalBottomSheetLayout extends SingleChildLayoutDelegate {
```
I might be wrong but given the iOS animation, I'm somewhat tempted to think that we might not need to use something this involved.

xster reviewed (commented):
Thanks again for your contribution. This is an important change. Thanks for putting this much work into it.

I took a high level look at the class relations themselves. 

My general feedback is that we may be able to avoid pushing so many things into the widgets layer. There are many new public classes created with very generic sounding names. In general, every API expansion in the framework should be very deliberate and should solve a general problem rather than being a connecting piece for one feature. 

On example of material/cupertino sharing is to look at the PageTransitionsTheme class where a lot of the implementation details are hidden behind the same existing page routes. I suspect we can keep a similar format. 

Also as @DanielEdrisian mentioned, we should make sure the curves are right for the animation. 

Lastly, we need to account for the navigator 2.0 APIs where @chunhtai may have some better instructions.

chunhtai commented:
This is implemented the same as showdialog, it will be treated as pageless route and should be fine unless we want to have decorative api to show the button sheet

jamesblasco commented:
Before going in detail answering back all the comments. Thanks, @xster for having a deep look at my code. 

In general, both `showMaterialModalBottomSheet` and `showCupertinoModalBottomSheet` API are heavily inspired in `showModalBottomSheet` (Notice it is different from `Scaffold.showBottomSheet`). 
`showCupertinoModalBottomSheet` would correspond to the Cupertino Modal View.
Also, I kept to showCustomModalBottomSheet to allow people to create their own easily 

eg: 
|  Example | Implementation |  
|---|---|
| <img height="200" src="https://iosimage.s3.amazonaws.com/2019/62-bottom-card/examples.png" />  https://fluffy.es/facebook-draggable-bottom-card-modal-1/ | https://github.com/jamesblasco/modal_bottom_sheet/blob/master/lib/src/bottom_sheets/bar_bottom_sheet.dart  | 
| <img height="200" src="https://user-images.githubusercontent.com/19904063/95793130-391a4880-0ce5-11eb-9df3-c49e56e0ad08.png" />   https://fluffy.es/facebook-draggable-bottom-card-modal-1/  | https://github.com/jamesblasco/modal_bottom_sheet/blob/master/example/lib/modals/circular_modal.dart  |  

All the implementations share most of the code and only the part for the decoration is separated for each. In the case of Cupertino, there is also a different route for the z-index translation.

If you consider we should only add the Cupertino Modal View with another API, I could work on that. I like the flexibility this current implementation has but I also understand why it would be good to focus on the Cupertino Modal View and remove the rest.

jamesblasco committed:
Improve readability and write types explicitly

jamesblasco committed:
Improve readability and write types explicitly

jamesblasco committed:
Fix typo

jamesblasco committed:
Improve comment

xster commented:
Regarding:

> This is implemented the same as showdialog, it will be treated as pageless route and should be fine unless we want to have decorative api to show the button sheet;

> In general, both showMaterialModalBottomSheet and showCupertinoModalBottomSheet API are heavily inspired in showModalBottomSheet 

Ya, both are very sensible. Also, I could be wrong about this given more data. I might not be making a complete observation here either. Since Apple's APIs are evolving as well, rather than freezing in Apple's current implementation, we have to map their APIs into our APIs' intents as well and some of which will be a subjective call. 

To me, the main differentiation we need to make between the modal pop-over style APIs we have such as `showModalBottomSheet`, `showCupertinoModalPopup` etc and the "permanent" page style APIs such as `PageRoute` with `CupertinoFullscreenDialogTransition` is how transient, non-resting-state the page in question are. 

Regarding https://fluffy.es/facebook-draggable-bottom-card-modal-1/ and the current dialogs, Cupertino action sheets etc, despite visual similarities, I believe they serve very different purposes than iOS 13's modal sheets. Consider for instance this sequence of actions: https://streamable.com/qxzonv

While modal (i.e. you need to stay here or in a child page until you complete the task), they're not lightweight, ephemeral or leaf nodes in the navigation. They need to deeply integrate with the rest of our PageRoutes (seen by the horizontal style transition in the middle of the existing vertical sheet transation) https://github.com/flutter/flutter/issues/33799. An equivalent question to ask is can a dialog itself be in the middle of a sequence of navigations that subsequently lead into another fullscreen page on top of it. 

Regarding navigator 2.0, another distinguishing question we can ask ourselves is could users deep link into one of the sheet presented pages and could a sheet presented page be serialized along with the rest of the navigation stack and then restored? I think the answer for both questions are no for a dialog box or a Cupertino action sheet and yes for a modal sheet.

jamesblasco commented:
Yes for me it is way different from how a dialog and a modal page should behave. I agree with you that the iOS modal is not as ephemeral as the `modal bottom sheet` described in the material guidelines.

iOS Sheet Modals should be able to display it with a deep link and allow inside-navigation. Related to the last one, I think users should use a nested `Navigator` or a  [`CupertinoTabView`](https://api.flutter.dev/flutter/cupertino/CupertinoTabView-class.html). I think in iOS is the same idea than with the `UINavigationController` ([ref](https://stackoverflow.com/questions/25444213/presenting-viewcontroller-with-navigationviewcontroller-swift)).

One concern is that I haven't seen a case where a different 'PageRoute' can push from a modal page. I don't think it is possible on iOS, but it would be possible on Flutter just by pushing a new route to the `Navigator`.  

I implemented the same API as the `showModalBottomSheet` as they have a lot of similarities, but I agree that should be exposed as a Route and not as a static API.

But for me, there is a blurred line between the material modal bottom sheet and the iOS modal sheet page.  There are some edge examples I have been researching:

<details>
<summary> Modal Sheets examples </summary>

1. **Apple Home App. Official iOS Modal Page.**
The same case happens on Spotify
Modal sheet don't animate the previous route
<img height="200" src="https://user-images.githubusercontent.com/19904063/95962937-534c4780-0e07-11eb-9a1d-da306c751433.PNG"> </img>

2.  **Apple Store App. Official iOS Modal Page. iPad Version**
The modal page on the iPad is displayed as dialog animated from the bottom. No code changes from the iPhone modal version.

<img height="200" src="https://user-images.githubusercontent.com/19904063/95967417-82b18300-0e0c-11eb-8a05-8f393869ae10.png"> </img>

3. **iOS Share Sheet**
Can be expanded as a full modal page. Can have nested modals. 

<img height="200" src="https://user-images.githubusercontent.com/19904063/95963471-04eb7880-0e08-11eb-9ae0-ff9f2ce45b08.PNG"> </img>

4.  **Material Modal Bottom Sheet **

The initial vertical position is capped at 50% of the screen height.(❌ Different)
Expandable full screen.
The concept is limited to simple actions. For other uses, material recommends [Expanding Bottom Sheet](https://material.io/components/sheets-bottom#expanding-bottom-sheet) that animate from a view instead of from the bottom.

<img height="200" src="https://user-images.githubusercontent.com/19904063/95967825-023f5200-0e0d-11eb-8032-019d1a473f61.png"> </img>

2. **Facebook App. Custom Modal Bottom Sheet.**

It has nested modals. And the models can have inside the navigation.  They are expanded or custom height. Could they be deep linked?

<img height="200" src="https://user-images.githubusercontent.com/19904063/95963117-8f7fa800-0e07-11eb-85f0-05198f5e2ff7.PNG"> </img>

4. **Slack Modal Bottom Sheet** https://github.com/slackhq/PanModal
They can be expanded, nested, and have inside navigation. They could also be deep linked.
 
<img height="200" src="https://github.com/slackhq/PanModal/raw/master/Screenshots/panModal.gif"> </img>
</details>

In the examples, for me, they share the same concepts and use different appearances to be displayed. All are animated from the bottom, can be closed by dragging, can be fit/expanded, and if there is a scroll view both scroll and drag should be sync. They can also be nested and used inside the navigation.

Maybe the naming is not the same but the implementation is. I understand that we should prioritize the material modal_bottom_sheet concept just for actions and use another name for the rest.

jamesblasco commented:
Sorry for the ping, @christopherfujino @Piinks. I made a rebase from master to my fork and somehow with Github Desktop, I added your lastest commits to this PR. Fixed now

xster commented:
Thanks for putting so much thought into it. It's definitely a gray area and the data you collected is helpful.

> Related to the last one, I think users should use a nested Navigator or a CupertinoTabView. On iOS, UINavigationController it is used with the same idea (ref).

Ya, at a glance, I suspect that'll have to be the route we take too. Though we don't have to solve https://github.com/flutter/flutter/issues/33799 the nesting problem right now in this PR, as long as we move in the right direction. 

> One of my concerns is that, on iOS, I haven't seen a case where a different 'PageRoute/UIViewController' can push from a modal page.

What do you mean? Is this related to https://github.com/flutter/flutter/issues/33799 or something else? Maybe you meant whether it's possible to push something not nested into the modal sheet but overtop it? If so I agree. 

> Apple Home App. Official iOS Modal Page.

Umm weird, it does animate the bottom page for me. Might be a bug.

> Apple Store App. Official iOS Modal Page. iPad Version

This is a great example. I was just thinking we might be able to make the optimization in case of modal sheets to turn the bottom route offscreen. This example definitely makes the case for it. While we don't currently support tablet formats (such as context menus instead of action sheets etc), we should definitely design architecturally with that in mind. 

> iOS Share Sheet

Great example too. I think the learning from that one is snapping at the 50% position will need to be something we leave room for as well. Also reinforces that the bottom route is still on-screen even for phones. 

> I understand that we should prioritize the material modal_bottom_sheet concept just for actions and use another name for the rest.

Ya, that's more or less in line with what I'm thinking

> So what about calling it CupertinoSheetRoute? And it would extend SheetRoute that can be used to display these kinds of pages that appear from the bottom?

So you're saying fork at the ModalRoute level and let there be a PageRoute, SheetRoute and PopupRoute as options? Without looking too deeply into the framework for existing assumptions and usages of PageRoute that are already baked into the framework, that sounds reasonable to me. @goderbauer might have some opinions on creating a branch there too (also wrt navigator 2.0).

xster commented:
@jamesblasco we love what you've done so far :) would you like to proceed forward with this PR? (or perhaps you're waiting for another review in its current state?)

jamesblasco commented:
I was waiting to see if the person you tagged would add something meaningful. 

But I can start to make the api as we said with SheetRoute.
This week I tried to move this to a project as it was easier for me to try things there, but I didn’t manage to build the project with the no-nullability code inside.

xster commented:
@goderbauer thoughts on adding SheetRoute as a sibling to PageRoute?

jamesblasco committed:
Refactor to Sheet Api

jamesblasco commented:
I've pushed new changes and the new API would now look like this:

```dart
Navigator.of(context).push(
    CupertinoSheetRoute(
       builder: (context) => Container(),
    ),
);
```

You can try some examples in this project: https://github.com/jamesblasco/modal_bottom_sheet_example/

For the next steps, I will try it with the new Navigator 2.0 and I will focus on the `Sheet` widget that is inside.

I want to add the option that it snaps to different positions.(See the iOS Share Sheet as example)  The approach will be similar to [DraggableScrollableSheet](https://api.flutter.dev/flutter/widgets/DraggableScrollableSheet-class.html), but with different behavior. While `DraggableScrollableSheet` you can drag from a min to max position, this widget should auto-snap to the min, max, or somewhere in the middle position. 

We can make this class `Sheet` private and use only it in SheetRoute or adapt it so it can be also used outside the route (like `DraggableScrollableSheet`). If the name is not descriptive enough we could be named it `SpanSheet`.

A thing that still need to be considered out: To sync a ScrollView with the Sheet you need to use the same scrollController:
Right now there is two options to do this:
```dart
CupertinoSheetRoute(
   builder: (context) => SingleChildScrollView(
      primary: true,
      child: Container(),
    ),
),
```
or 
```dart
CupertinoSheetRoute(
   builder: (context) => SingleChildScrollView(
      controller: SheetController.of(context).scrollController,
      child: Container(),
    ),
),
```
I prefer the first one but I think people will have problems with it because of this: https://github.com/flutter/flutter/issues/64236. Using Scaffold inside the SheetRoute would make this non-usable, eg:
```dart
CupertinoSheetRoute(
   builder: (context) => Scaffold(
     body: Builder(
       builder: (context) => SingleChildScrollView(
           primary: true,
           child: Container(),
        ),
     ),
   ),
),
```

jamesblasco commented on packages/flutter/lib/src/widgets/sheet_route.dart:
```diff
@@ -0,0 +1,266 @@
+
+/// Route that defines the transition animation of the previous route when this
+/// one is closing
+mixin RouteWithPreviousTransitionMixin<T> on Route<T> {
+  Widget getPreviousRouteTransition(
+    BuildContext context,
+    Animation<double> secondAnimation,
+    Widget child,
+  );
+}
+
```
I think this would be a good way to start. I could make a different PR with this code to allow back transitions to be defined from the next Route.

Right now it consists of two mixins: 
`AllowsRouteWithPreviousTransitionMixin` -> Assigned to the route that will transition with the custom animation eg. MaterialPageRoute
`RouteWithPreviousTransitionMixin` -> Assigned to the route that will be pushed and will define the transition for the previous route eg. CupertinoSheetRoute.

If you agree we can move this to another PR and I will start working on the tests and docs.

xster commented on packages/flutter/lib/src/widgets/sheet_route.dart:
```diff
@@ -0,0 +1,266 @@
+        final Animation<double> fakeSecondaryAnimation =
+            Tween<double>(begin: 0, end: 0).animate(secondaryAnimation);
+
+        final Widget defaultTransition = super.buildTransitions(
```
I see your dilemma (in that we never had a pattern to let the bottom route transition different depending on the top route except CupertinoFullscreenDialogTransition which "cheats" by not transitioning the bottom route at all). Though I think we should solve this problem more thoroughly. I see 2 general approaches. Either we create a general plumbing (maybe somewhere here https://github.com/flutter/flutter/blob/b44696626ce329d899f5c29d7664c35b89caad06/packages/flutter/lib/src/widgets/routes.dart#L802) to always let the top route inject a transition to override the transition of the bottom route or we can create something specific for Cupertino and just let CupertinoRouteTransitionMixin be a mixing on ModalRoute instead of PageRoute, then just let it explicitly know and orchestrate between CupertinoFullscreenDialogTransition, CupertinoPageTransition and a new CupertinoSheetTransition (with a bit more work to build in handles to handle all 3 possible top route types in the bottom transition)

goderbauer replied:
+1 to finding a more generic way of doing this. I think that should be based in the widget layer, not just cupertino.

jamesblasco replied:
I am investigating and `TransitionRoute` uses an approach similar to this when a route returns false from canTransitionTo/canTransitionFrom.

The transition route will use `kAlwaysDismissedAnimation` as secondaryAnimation in this case

https://github.com/flutter/flutter/blob/84f3d28555368a70270e9ac8390a9441df95e752/packages/flutter/lib/src/widgets/routes.dart#L344
https://github.com/flutter/flutter/blob/84f3d28555368a70270e9ac8390a9441df95e752/packages/flutter/lib/src/animation/animations.dart#L74-L79

xster commented on packages/flutter/lib/src/widgets/sheet_route.dart:
```diff
@@ -0,0 +1,266 @@
+
+const Duration _bottomSheetDuration = Duration(milliseconds: 400);
+
+class _Sheet<T> extends StatefulWidget {
```
I'm a bit surprised there's a _Sheet and _SheetState here and another Sheet and _SheetState in the other file. Maybe the names need to be more specific about what they do.

xster reviewed (commented):
Thanks for making the change. 

For the widget layer changes, I didn't review all the details yet but just the APIs so far. I think the main piece that might shake things up a bit to need another round is around the nesting of the transition building to handle the previous route. Generally, don't be afraid to change the existing transition mixins to account for the new type directly rather than blackboxing them and then entirely wrapping them with a fake secondaryAnimation.

goderbauer commented:
> @goderbauer thoughts on adding SheetRoute as a sibling to PageRoute?

Sorry for the delay in looking at this. It's the first thing in my queue now and I will take a look at this first thing tomorrow morning.

goderbauer commented on packages/flutter/lib/src/widgets/sheet.dart:
```diff
@@ -0,0 +1,677 @@
+///
+///  * [showMaterialModalBottomSheet] which can be used to display a modal bottom
+///    sheet with Material appareance.
+///  * [showCupertinoModalBottomSheet] which can be used to display a modal bottom
```
I can't find these anywhere. How do I use the sheet API as a developer?

jamesblasco replied:
This is removed now, it was part of the old API I use in the original package 
https://github.com/jamesblasco/modal_bottom_sheet

goderbauer commented on packages/flutter/lib/src/widgets/sheet.dart:
```diff
@@ -0,0 +1,677 @@
+import 'package:flutter/scheduler.dart';
+import 'package:flutter/widgets.dart';
+
+const Curve _decelerateEasing = Cubic(0.0, 0.0, 0.2, 1.0);
```
I am surprised that the widget layer has so much opinion on the visuals of the sheet. Shouldn't this be upto the design languages to configure?

goderbauer commented:
> So you're saying fork at the ModalRoute level and let there be a PageRoute, SheetRoute and PopupRoute as options? Without looking too deeply into the framework for existing assumptions and usages of PageRoute that are already baked into the framework, that sounds reasonable to me. @goderbauer might have some opinions on creating a branch there too (also wrt navigator 2.0).

How do `PageRoute` and `SheetRoute` interact with each other? My understanding from iOS is that once you have pushed a sheet, you can push other routes within that sheet as well. What are those routes within the sheet? Do I need a separate Navigator for those in the sheet and then push regular PageRoutes onto them? Can I just push other `SheetRoute`s and `PageRoute`s and they will show up within the sheet? While on a sheet can I push another sheet? I think to decide whether a SheetRoute makes sense as a parallel concept to PageRoute I need to better understand how those two interact with each other.

These questions also extend to Navigator 2.0. If I have a CupertinoSheetPage in my list of pages and the next page is a CupertinoPage, will it fully cover the SheetPage? Will it somehow show up within the SheetPage?

As general feedback: It would be good to make it a little clearer what the entry point for the API is. Reading through the code it wasn't clear to me what is part of the public API and what isn't really. Better documentation would also help with figuring out how the pieces fit together.

goderbauer reviewed (commented):
(not a full review, just some random comments while scrolling through the code)

goderbauer commented:
Ok, After re-reading the comment thread and talking to @xster I have a better understanding of this subject:

I agree that the iOS sheet should behave more like a route and less like a modal. It is more permanent than a modal and it may include other "screens" (see for example the calendar editing flow). Given that, I think there should be something you push actively on the Navigator to show the sheet. This could be a special SheetRoute (which my include a sub-navigator to present other PageRoutes within the sheet - we should consider whether certain SheetRoute types always add a sub-navigator to achieve this since it seems pretty common and I think it doesn't make much sense to push a full-screen PageRoute over a SheetRoute?). 

The more interesting aspect is still how PageRoutes and SheetRoutes will interact with each other. The most difficult part here seems to be that the new route needs to tell the old route how to transition off screen. That's something our routing system doesn't currently support, AFAIK. It may make sense to design this mechanism first. I think it should be a general mechanism of Routes, that's not specific to cupertino routes. 

So, to answer the original question:

> So you're saying fork at the ModalRoute level and let there be a PageRoute, SheetRoute and PopupRoute as options? Without looking too deeply into the framework for existing assumptions and usages of PageRoute that are already baked into the framework, that sounds reasonable to me. @goderbauer might have some opinions on creating a branch there too (also wrt navigator 2.0).

I am not opposed to adding a new SheetRoute route type.

xster commented:
@goderbauer's comment sound reasonable to me. @jamesblasco, I think the central comment in our latest round is around https://github.com/flutter/flutter/pull/67515#discussion_r512271702. Generally, don't feel constrained by the current navigator implementation and feel free to shake up our transition building mechanism (see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#lazy-programming). 

Since that design is a bit more open-ended, feel free to either write a short doc (https://flutter.dev/docs/resources/design-docs) or continue to write proposal code in your PR branch to discuss the high level design. Whichever you're more comfortable with. We're happy to give you more feedback.

jamesblasco committed:
Implement SnapSheet and add comments

jamesblasco commented:
Thanks for all the comments both @xster and @goderbauer. I will take a detailed look at all of them one by one and fix the respective code.

I have pushed a new version that I summary a lot of the things we have talked about.

The public classes would be:
- `SnapSheet` a widget similar to `DraggableScrollableSheet` that can snap in different anchors.
You can control it with `SheetController`( if we keep the name `SnapSheet` it should be named `SnapSheetControlller`) 
and use it anywhere in your app. I am not very confident about this class logic.
eg:

<img  height="400" src="https://user-images.githubusercontent.com/19904063/97347558-40b52200-188d-11eb-9ca1-58a3c98f2d22.gif" />

- Then there is `SheetRoute` that uses `SnapSheet` internally. `CupertinoSheetRoute` is an implementation of `SheetRoute`
that also animates the previous/bottom route.

<img  height="400" src="https://user-images.githubusercontent.com/19904063/97347864-af927b00-188d-11eb-832b-ca76e6dec294.gif" />

I've also mplemented an example with Navigator 2.0. I recommend you to run the examples repository I have created https://github.com/jamesblasco/modal_bottom_sheet_example

This should be the only public classes available after this PR.

For me the next steps would be:

1. Create another PR or design doc for the following: Routes that tell the old route how to transition off-screen.

2. Fix animation curves and todos. I left some todos in the code about related I am working now

3. Start adding test, if the part from above is clear I should start now with this.

4. Improve docs and follow 100% style guidelines.

jamesblasco committed:
Add license headers and local imports

jamesblasco commented on packages/flutter/lib/src/cupertino/sheet_route.dart:
```diff
@@ -0,0 +1,237 @@
+
+/// Estimated Round corners for iPhone X, XR, 11, 11 Pro
+/// https://kylebashour.com/posts/finding-the-real-iphone-x-corner-radius
+const double _kDeviceFrameCorners = 38.5;
```
About this, I saw that in the link from the doc, the dev has updated the article and he says there is a private API from Apple that tells you which corner radius has the device.

<img width="475" alt="Captura de pantalla 2020-10-27 a las 12 30 33" src="https://user-images.githubusercontent.com/19904063/97351518-2b8ec200-1892-11eb-86ea-631e60366fa2.png">

This means there are multiple radius for different devices. In the iOS Modal Sheet, the bottom/previous page is animated to match that radius.

Is there any way we could know which device is it and set a radius for each?

jamesblasco commented on packages/flutter/lib/src/cupertino/sheet_route.dart:
```diff
@@ -0,0 +1,237 @@
+    Curve? animationCurve,
+    Color? modalBarrierColor,
+    bool bounce = true,
+    bool isDismissible = true,
```
Mmm if the sheet is not fully expandable then is dismissible. With the current ModalBarrier, can we make this change depending on the `animationController`?

jamesblasco commented on packages/flutter/lib/src/cupertino/sheet_route.dart:
```diff
@@ -0,0 +1,237 @@
+    bool bounce = true,
+    bool isDismissible = true,
+    bool enableDrag = true,
+    bool expanded = true,
```
Yes, that is what it means, and no there is no equivalent that I know.

jamesblasco commented on packages/flutter/lib/src/cupertino/sheet_route.dart:
```diff
@@ -0,0 +1,237 @@
+    bool expanded = true,
+    Duration? duration,
+    RouteSettings? settings,
+    this.transitionBackgroundColor,
```
Yes is the background behind all the sheets. In the package, people ask me to add it for customization.
iOS doesn't allow to customize, so I have removed it

jamesblasco commented on packages/flutter/lib/src/cupertino/sheet_route.dart:
```diff
@@ -0,0 +1,237 @@
+  ) {
+    final double paddingTop = MediaQuery.of(context)?.padding.top ?? 0;
+    return AnimatedBuilder(
+      animation: secondaryAnimation,
```
This is used to animate a CupertinoSheetRoute when the next route is a CupertinoSheetRoute. But like it is now it would always animate when the next route is a SheetRoute, should fix that

jamesblasco commented on packages/flutter/lib/src/cupertino/sheet_route.dart:
```diff
@@ -0,0 +1,237 @@
+    final double paddingTop = MediaQuery.of(context)?.padding.top ?? 0;
+
+    final bool phoneWithRoundedCorners =
+        defaultTargetPlatform == TargetPlatform.iOS && paddingTop > 20;
```
I only measured the iPhone X from the iOS HIG sketch file. This would need a better research

jamesblasco committed:
Fix typo

xster commented:
@willlarche also called attention to this new Material spec https://standards.google/guidelines/google-material/using-google-material/gm-for-ios/#adaptations. Thanks for pointing it out. The overall design implication for us is that it's worth considering making the new route type accessible from the Material stack as well and thinking about how to integrate it with the Material modal APIs. 

Thanks for writing the doc @jamesblasco. Will review tonight.

jamesblasco commented:
@xster I am not sure what you are talking about. I don't have access to the link you shared. What is the new Material spec? Is it public or internal for google?

xster commented:
Sorry, ignore my last comment. It wasn't correct.

gibahjoe commented:
Bump!!!

sanjade commented:
Any progress?

sanjade commented:
@jamesblasco 👀

SAGARSURI commented:
Any progress??

tvolkert commented:
@xster can you recommend a good person to pick up this review?

@jamesblasco it looks like there are some failing tests - can you please rebase and fix the tests?

jamesblasco commented:
Hello @tvolkert, I am keeping working on improving the modals sheet through other channels. This PR is blocked by #69669. I haven't heard back there for a while

Hixie commented:
@jamesblasco Well it's more than a year after you submitted a PR so I apologize about the absurd slowness of this review but I finally got around to looking at the samples and the code and wow, this is really cool. I will look more closely at the code now but at a glance you seem to be doing all the right things.

Hixie commented on packages/flutter/lib/src/cupertino/sheet_route.dart:
```diff
@@ -0,0 +1,254 @@
+            scale: scale,
+            child: child,
+            alignment: Alignment.topCenter,
+          ),
```
it might be possible to use something like SlideTransition and ScaleTransition rather than an AnimatedBuilder, in principle those will be faster because they can skip the rebuild/relayout and only repaint.

Hixie replied:
(they don't currently, but the point is that if we improve them later, it'd improve this code too)

Hixie commented on packages/flutter/lib/src/cupertino/sheet_route.dart:
```diff
@@ -0,0 +1,254 @@
+    final double topOffset = math.max(_kSheetMinimalOffset, topPadding);
+
+    final bool isRoundedDevice =
+        defaultTargetPlatform == TargetPlatform.iOS && topPadding > _kRoundedDeviceStatusBarHeight;
```
nit: avoid `==` with enums, instead use `switch`. That way if we add a new TargetPlatform the linter will flag this code as needing consideration.

Hixie replied:
is there a more reliable way of determining if it's a "rounded device"? e.g. is the recent MacBook Pro considered "rounded"? (I guess not because the roundness is where the menu bar goes?)

jamesblasco commented:
Hi @Hixie, I am not working on this PR anymore. I don't think it is a good investment of your time right now to review this

Hixie commented on packages/flutter/lib/src/widgets/pages.dart:
```diff
@@ -116,3 +116,71 @@ class PageRouteBuilder<T> extends PageRoute<T> {
+
+// TODO(jamesblasco): Rethink this to no use extra mixin or a fakeSecondaryAnimation 
+/// Route that defines the transition animation of the previous route when this
+/// one is closing/opening
```
yeah we may just need to rethink how we do transitions in this way in general

Hixie commented on packages/flutter/lib/src/widgets/sheet.dart:
```diff
@@ -0,0 +1,1036 @@
+const Duration _kDefaultSheetDuration = Duration(milliseconds: 400);
+const Duration _kBounceAnimationDuration = Duration(milliseconds: 300);
+
+/// NEEDS DOCS
```
minor hint: if you just skip the docs entirely, the linter will tell you where docs are missing (but when you put placeholder docs with `///`, it won't)

Hixie commented on packages/flutter/lib/src/widgets/sheet.dart:
```diff
@@ -0,0 +1,1036 @@
+    _controller = widget.controller ?? SheetController(vsync: this);
+  }
+
+  // TODO(jamesblasco): is there any way to access the Drag inside a GestureDetector and cancel it?
```
not using GestureDetector but you can go one level deeper and create the `Drag` object manually and then manage it that way, iirc.

Hixie commented:
@jamesblasco sounds good. sorry again that we dropped the ball on this.
